File: app/src/main/java/ru/p3tr0vich/fuel/FragmentDialogProgress.java
package ru.p3tr0vich.fuel;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AlertDialog;
import android.view.View;
import android.widget.TextView;

import ru.p3tr0vich.fuel.helpers.DatabaseBackupXmlHelper;
import ru.p3tr0vich.fuel.utils.UtilsLog;

public class FragmentDialogProgress extends DialogFragment {

    public static final String TAG = "FragmentDialogProgress";

    private static final String MESSAGE = "message";
    private static final String EXTRA_XML_RESULT = "EXTRA_XML_RESULT";

    private AsyncTaskOperationXml mAsyncTaskOperationXml;

    public static void show(@NonNull Fragment parent, int requestCode,
                            @NonNull DatabaseBackupXmlHelper databaseBackupXmlHelper,
                            boolean doSave) {
        FragmentDialogProgress fragmentDialogProgress = new FragmentDialogProgress();

        Context context = parent.getContext();

        assert context != null;

        fragmentDialogProgress.setTask(new AsyncTaskOperationXml(context, databaseBackupXmlHelper, doSave));
        fragmentDialogProgress.setTargetFragment(parent, requestCode);

        Bundle args = new Bundle();
        args.putString(MESSAGE, parent.getString(doSave ? R.string.message_progress_save : R.string.message_progress_load));
        fragmentDialogProgress.setArguments(args);

        FragmentManager fragmentManager = parent.getFragmentManager();

        assert fragmentManager != null;

        fragmentDialogProgress.show(fragmentManager, TAG);
    }

    @DatabaseBackupXmlHelper.BackupResult
    public static int getResult(Intent data) {
        return DatabaseBackupXmlHelper.intToResult(data.getIntExtra(EXTRA_XML_RESULT, -1));
    }

    private void setTask(AsyncTaskOperationXml asyncTaskOperationXml) {
        UtilsLog.d(TAG, "setTask");

        mAsyncTaskOperationXml = asyncTaskOperationXml;
        mAsyncTaskOperationXml.setFragmentDialogProgress(this);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setCancelable(false);
        setRetainInstance(true);

        UtilsLog.d(TAG, "onCreate");

        if (mAsyncTaskOperationXml != null) mAsyncTaskOperationXml.execute();
    }

    @Override
    public void onResume() {
        super.onResume();

        UtilsLog.d(TAG, "onResume",
                "mAsyncTaskOperationXml " + (mAsyncTaskOperationXml == null ? "=" : "!") + "= null");

        if (mAsyncTaskOperationXml == null) dismiss();
    }

    @Override
    public void dismiss() {
        UtilsLog.d(TAG, "dismiss");
        super.dismiss();
    }

    @Override
    public void onDestroyView() {
        if (getDialog() != null && getRetainInstance())
            getDialog().setDismissMessage(null);
        super.onDestroyView();
    }

    void stopTask(@DatabaseBackupXmlHelper.BackupResult int result) {
        UtilsLog.d(TAG, "stopTask",
                "getTargetFragment() " + (getTargetFragment() == null ? "=" : "!") + "= null");

        if (isResumed()) {
            UtilsLog.d(TAG, "stopTask", "isResumed()");
            dismiss();
        }

        mAsyncTaskOperationXml = null;

        if (getTargetFragment() != null)
            getTargetFragment().onActivityResult(getTargetRequestCode(), Activity.RESULT_OK,
                    new Intent().putExtra(EXTRA_XML_RESULT, result));
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        UtilsLog.d(TAG, "onCreateDialog");

        Activity activity = getActivity();

        assert activity != null;

        @SuppressLint("InflateParams")
        View rootView = activity.getLayoutInflater().inflate(R.layout.dialog_progress, null, false);

        Bundle bundle = getArguments();

        assert bundle != null;

        ((TextView) rootView.findViewById(R.id.text_message)).setText(bundle.getString(MESSAGE));

        return new AlertDialog.Builder(activity).setView(rootView).create();
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/fragments/FragmentDialogProgress.kt
package ru.p3tr0vich.fuel.fragments

import android.annotation.SuppressLint
import android.app.Activity
import android.app.Dialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import ru.p3tr0vich.fuel.AsyncTaskOperationXml
import ru.p3tr0vich.fuel.R
import ru.p3tr0vich.fuel.helpers.DatabaseBackupXmlHelper
import ru.p3tr0vich.fuel.utils.UtilsLog

class FragmentDialogProgress : DialogFragment() {

    private var asyncTaskOperationXml: AsyncTaskOperationXml? = null
        set(value) {
            UtilsLog.d(TAG, "setTask == $value")

            field = value

            field?.fragmentDialogProgress = this
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        isCancelable = false
        retainInstance = true

        UtilsLog.d(TAG, "onCreate", "asyncTaskOperationXml == $asyncTaskOperationXml")

        asyncTaskOperationXml?.execute()
    }

    override fun onResume() {
        super.onResume()

        UtilsLog.d(TAG, "onResume", "asyncTaskOperationXml == $asyncTaskOperationXml")

        if (asyncTaskOperationXml == null) {
            dismiss()
        }
    }

    override fun dismiss() {
        UtilsLog.d(TAG, "dismiss")

        super.dismiss()
    }

/*    override fun onDestroyView() {
        if (dialog != null && retainInstance) {
// TODO:           dialog.setDismissMessage(null)
        }

        super.onDestroyView()
    }*/

    internal fun stopTask(@DatabaseBackupXmlHelper.BackupResult result: Int) {
        UtilsLog.d(TAG, "stopTask", "targetFragment == $targetFragment")

        if (isResumed) {
            UtilsLog.d(TAG, "stopTask", "isResumed")
            dismiss()
        }

        asyncTaskOperationXml = null

        targetFragment?.onActivityResult(targetRequestCode, Activity.RESULT_OK,
                Intent().putExtra(EXTRA_XML_RESULT, result))
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        UtilsLog.d(TAG, "onCreateDialog")

        @SuppressLint("InflateParams")
        val rootView = requireActivity().layoutInflater.inflate(R.layout.dialog_progress, null, false)

        (rootView.findViewById<View>(R.id.text_message) as TextView).text = arguments?.getString(MESSAGE)

        return AlertDialog.Builder(requireActivity()).setView(rootView).create()
    }

    companion object {

        const val TAG = "FragmentDialogProgress"

        private const val MESSAGE = "message"
        private const val EXTRA_XML_RESULT = "EXTRA_XML_RESULT"

        fun show(parent: Fragment, requestCode: Int,
                 databaseBackupXmlHelper: DatabaseBackupXmlHelper,
                 doSave: Boolean) {
            val fragmentDialogProgress = FragmentDialogProgress()

            fragmentDialogProgress.asyncTaskOperationXml = AsyncTaskOperationXml(parent.requireContext(), databaseBackupXmlHelper, doSave)
            fragmentDialogProgress.setTargetFragment(parent, requestCode)

            val args = Bundle()
            args.putString(MESSAGE, parent.getString(if (doSave) R.string.message_progress_save else R.string.message_progress_load))
            fragmentDialogProgress.arguments = args

            fragmentDialogProgress.show(parent.parentFragmentManager, TAG)
        }

        @DatabaseBackupXmlHelper.BackupResult
        fun getResult(data: Intent): Int {
            return DatabaseBackupXmlHelper.intToResult(data.getIntExtra(EXTRA_XML_RESULT, -1))
        }
    }
}